[inductor] Implement magic methods on IR values

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101076

This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

peterbell10 committed:
[inductor] Fix invalid ops calls in upsample_bicubic2d lowering

These `cubic_convolution` helpers use `ops` calls with python constants instead
of calling `ops.constant` first to get an IR value.

I also introduce `OpsWrapper` as a way to write mathematical operations more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(A + 2, x), A + 3), x), x), 1.0)
```
we can write
```python
x = OpsWrapper(x)
_Ap2, _Ap3, _1 = OpsWrapper.constants(A + 2, A + 3, 1, dtype=torch.float32)
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Fix invalid ops calls in upsample_bicubic2d lowering"


These `cubic_convolution` helpers use `ops` calls with python constants instead
of calling `ops.constant` first to get an IR value.

I also introduce `OpsWrapper` as a way to write mathematical operations more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(A + 2, x), A + 3), x), x), 1.0)
```
we can write
```python
x = OpsWrapper(x)
_Ap2, _Ap3, _1 = OpsWrapper.constants(A + 2, A + 3, 1, dtype=torch.float32)
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Fix invalid ops calls in upsample_bicubic2d lowering"


These `cubic_convolution` helpers use `ops` calls with python constants instead
of calling `ops.constant` first to get an IR value.

I also introduce `OpsWrapper` as a way to write mathematical operations more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(A + 2, x), A + 3), x), x), 1.0)
```
we can write
```python
x = OpsWrapper(x)
_Ap2, _Ap3, _1 = OpsWrapper.constants(A + 2, A + 3, 1, dtype=torch.float32)
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps the underlying ops
handler's return type into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

peterbell10 committed:
lint on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

lezcano commented on torch/_inductor/virtualized.py:
```diff
@@ -153,21 +154,87 @@ def __getattr__(self, item):
 
 
+class OpsValue:
+    """The turn type of most ops calls. This exists so we can do operator
```
nit return

lezcano reviewed (approved):
Cool. We'll also be able to use this class in the future to track things that can be known at lowering time.

peterbell10 committed:
Update on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

ngimel reviewed (approved):
Nice! 
PR description should leave somewhere in the comments (because comments are forever, and PR description has to be searched).

peterbell10 committed:
Update on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Implement magic methods on IR values"


This wraps `ops` into an `OpsWrapper` object which wraps any returned
IR values into an `OpsValue` instance. This allows magic methods to
be implemented and means lowerings can write mathematical expressions much more
fluently. So instead of
```python
ops.add(ops.mul(ops.mul(ops.sub(ops.mul(_Ap2, x), _Ap3), x), x), _1)
```
we can write
```python
(_Ap2 * x - _Ap3) * x * x + _1
```

And it will translate to the equivalent `ops` calls.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge
